File: src/main/java/com/example/fileservice/controller/FileController.java
package com.example.fileservice.controller;

import com.example.fileservice.dto.FileUploadRequest;
import com.example.fileservice.controller.exception.InternalException;
import com.example.fileservice.controller.exception.NotFoundException;
import com.example.fileservice.data.entities.Entity;
import com.example.fileservice.service.FileService;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.validation.Valid;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/file")
@Validated
public class FileController {

    private final FileService fileService;
    private final ObjectMapper objectMapper;

    public FileController(FileService fileService, ObjectMapper objectMapper) {
        this.fileService = fileService;
        this.objectMapper = objectMapper;
    }

    @PostMapping("/upload")
    public ResponseEntity<Map<String, String>> uploadFile(@Valid @ModelAttribute FileUploadRequest request) {
        try {
            Map<String, Object> metaMap = objectMapper.readValue(request.getMeta(), new TypeReference<>() {
            });
            String token = fileService.uploadFile(
                    request.getName(),
                    request.getContentType(),
                    metaMap,
                    request.getSource(),
                    request.getExpireTime(),
                    request.getContent()
            );

            Map<String, String> response = Collections.singletonMap("token", token);
            return ResponseEntity.status(HttpStatus.CREATED).body(response);
        } catch (IOException e) {
            throw new IllegalArgumentException("Invalid metadata: " + e.getMessage(), e);
        } catch (Exception e) {
            throw new InternalException("Internal server error: " + e.getMessage(), e);
        }
    }

    @PostMapping("/metas")
    public ResponseEntity<Map<String, Object>> getFileMetadata(@RequestBody Map<String, List<String>> request) {
        try {
            List<String> tokens = request.get("tokens");
            if (tokens == null || tokens.isEmpty()) {
                throw new IllegalArgumentException("Tokens list cannot be null or empty");
            }

            List<UUID> uuids = tokens.stream()
                    .map(UUID::fromString)
                    .collect(Collectors.toList());
            Map<String, Entity> metadataMap = fileService.getMetadataByTokens(uuids);

            Map<String, Object> response = metadataMap.entrySet()
                    .stream()
                    .collect(Collectors.toMap(
                            Map.Entry::getKey,
                            entry -> Map.of(
                                    "token", entry.getKey(),
                                    "filename", entry.getValue().getName(),
                                    "size", entry.getValue().getSize(),
                                    "contentType", entry.getValue().getContentType(),
                                    "createTime", entry.getValue().getCreateTime().toInstant().toString(),
                                    "meta", entry.getValue().getMeta()
                            )
                    ));

            return ResponseEntity.ok(Map.of("files", response));
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException(e.getMessage(), e);
        } catch (Exception e) {
            throw new InternalException("Internal server error: " + e.getMessage(), e);
        }
    }

    @GetMapping("/{token}")
    public ResponseEntity<Resource> downloadFile(@PathVariable String token) {
        try {
            File file = fileService.getFile(token);
            if (file == null) {
                throw new NotFoundException("File not found for token: " + token);
            }

            Path path = file.toPath();
            Resource resource = new UrlResource(path.toUri());
            if (!resource.exists() || !resource.isReadable()) {
                throw new NotFoundException("File not readable for token: " + token);
            }

            Entity metadata = fileService.getMetadata(UUID.fromString(token));
            return ResponseEntity.ok()
                    .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + file.getName() + "\"")
                    .header("X-Filename", file.getName())
                    .header("X-Filesize", String.valueOf(file.length()))
                    .header("X-CreateTime", metadata.getCreateTime().toInstant().toString())
                    .body(resource);
        } catch (Exception e) {
            throw new InternalException("Internal server error: " + e.getMessage(), e);
        }
    }

    @DeleteMapping("/{token}")
    public ResponseEntity<Void> deleteFile(@PathVariable UUID token) {
        try {
            boolean isDeleted = fileService.deleteFileByToken(token);
            if (!isDeleted) {
                throw new NotFoundException("File not found for token: " + token);
            }
            return ResponseEntity.noContent().build();
        } catch (Exception e) {
            throw new InternalException("Internal server error: " + e.getMessage(), e);
        }
    }
}

File: src/main/kotlin/com/example/fileservice/controller/FileController.kt
package com.example.fileservice.controller

import com.example.fileservice.dto.FileUploadRequest
import com.example.fileservice.controller.exception.InternalException
import com.example.fileservice.controller.exception.NotFoundException
import com.example.fileservice.data.dto.FileUploadRequest
import com.example.fileservice.rest.ErrorMessage
import com.example.fileservice.service.FileService
import com.fasterxml.jackson.core.type.TypeReference
import com.example.fileservice.data.entities.Entity
import com.example.fileservice.service.FileService
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.validation.Valid
import org.springframework.core.io.Resource
import org.springframework.core.io.UrlResource
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import java.io.File
import java.io.IOException
import java.nio.file.Path
import java.util.*
import javax.validation.Valid


@RestController
@RequestMapping("/api/file")
class FileController(
    private val objectMapper: ObjectMapper,
    private val fileService: FileService
) {

    @PostMapping("/upload")
    fun uploadFile(@Valid @ModelAttribute request: FileUploadRequest): ResponseEntity<Map<String, String>> {
        return try {
            val metaMap: Map<String, Any> =
                objectMapper.readValue(request.meta, object : TypeReference<Map<String, Any>>() {})

            val token = fileService.uploadFile(
                request.name,
                request.contentType,
                metaMap,
                request.source,
                request.expireTime!!,
                request.content
            )

            val response = mapOf("token" to token)
            ResponseEntity.status(HttpStatus.CREATED).body(response)
        } catch (e: IOException) {
            throw IllegalArgumentException("${e.message}", e)
        } catch (e: IllegalArgumentException) {
            throw e
        } catch (e: Exception) {
            throw InternalException("Internal server error: ${e.message}", e)
        }
    }

    @PostMapping("/metas")
    fun getFileMetadata(@RequestBody request: Map<String, List<String>>): ResponseEntity<Map<String, Any>> {
        return try {
            val tokens = request["tokens"]
                ?: throw IllegalArgumentException("Tokens list cannot be null or empty")

            if (tokens.isEmpty()) {
                throw IllegalArgumentException("Tokens list cannot be null or empty")
            }

            val uuids = tokens.map { UUID.fromString(it) }
            val metadataMap = fileService.getMetadataByTokens(uuids)

            val response = metadataMap.entries.associate { (key, value) ->
                key to mapOf(
                    "token" to key,
                    "filename" to value.name,
                    "size" to value.size,
                    "contentType" to value.contentType,
                    "createTime" to value.createTime.toInstant().toString(),
                    "meta" to value.meta
                )
            }
            ResponseEntity.ok(mapOf("files" to response))
        } catch (e: IllegalArgumentException) {
            throw e
        } catch (e: Exception) {
            throw InternalException("Internal server error: ${e.message}", e)
        }
    }

    @GetMapping("/{token}")
    fun downloadFile(@PathVariable token: String): ResponseEntity<*> {
        return try {
            val file = fileService.getFile(token)
            file ?: throw NotFoundException("File not found for token: $token")

            val path = file.toPath()
            val resource: Resource = UrlResource(path.toUri())
            if (!resource.exists() || !resource.isReadable) {
                throw NotFoundException("File not readable for token: $token")
            }

            val metadata = fileService.getMetadata(UUID.fromString(token))
            ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"${file.name}\"")
                .header("X-Filename", file.name)
                .header("X-Filesize", file.length().toString())
                .header("X-CreateTime", metadata.createTime.toInstant().toString())
                .body(resource)
        } catch (e: NotFoundException) {
            ResponseEntity.status(HttpStatus.NOT_FOUND).body(ErrorMessage(e.message))
        } catch (e: Exception) {
            throw InternalException("Internal server error: ${e.message}", e)
        }
    }

    @DeleteMapping("/{token}")
    fun deleteFile(@PathVariable token: UUID): ResponseEntity<ErrorMessage> {
        return try {
            val isDeleted = fileService.deleteFileByToken(token)
            if (!isDeleted) {
                throw NotFoundException("File not found for token: $token")
            }
            ResponseEntity.noContent().build()
        } catch (e: NotFoundException) {
            ResponseEntity.status(HttpStatus.NOT_FOUND).body(ErrorMessage(e.message))
        } catch (e: Exception) {
            throw InternalException("Internal server error: ${e.message}", e)
        }
    }
}
